sources/api/thar-be-updates/src/status.rs

```diff
@@ -0,0 +1,391 @@
+        &self.version
+    }
+}
+
+/// StagedImage represents a Bottlerocket image that is written to a partition set
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct StagedImage {
+    image: UpdateImage,
+    // Indicates whether this image is marked for next boot
+    next_to_boot: bool,
+}
+
+impl StagedImage {
+    pub(crate) fn set_next_to_boot(&mut self, next_to_boot: bool) {
+        self.next_to_boot = next_to_boot
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum CommandStatus {
+    Success,
+    Failed,
+    Unknown,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum UpdateCommand {
+    RefreshUpdate,
+    PrepareUpdate,
+    ActivateUpdate,
+    DeactivateUpdate,
+}
+
+/// CommandResult represents the result of an issued command
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct CommandResult {
+    cmd_type: UpdateCommand,
+    cmd_status: CommandStatus,
+    timestamp: DateTime<Utc>,
+    exit_status: Option<i32>,
+    stderr: Option<String>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct UpdateStatus<T> {
+    update_state: UpdateState,
+    available_updates: Vec<T>,
+    chosen_update: Option<UpdateImage>,
+    active_partition: Option<StagedImage>,
+    staging_partition: Option<StagedImage>,
+    most_recent_command: Option<CommandResult>,
+}
+
+impl Default for UpdateStatus<update_metadata::Update> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Loads and returns the update status from disk.
+pub fn get_update_status() -> Result<UpdateStatus<update_metadata::Update>> {
+    let status_file = File::open(UPDATE_STATUS_FILE).context(error::OpenStatusFile {
+        path: UPDATE_STATUS_FILE,
+    })?;
+    Ok(serde_json::from_reader(status_file).context(error::StatusParse)?)
+}
+
+// This is how the UpdateStatus is stored on disk
+impl UpdateStatus<update_metadata::Update> {
+    /// Initializes the update status
+    pub fn new() -> Self {
+        Self {
+            update_state: UpdateState::Idle,
+            available_updates: vec![],
+            chosen_update: None,
+            active_partition: None,
+            staging_partition: None,
+            most_recent_command: None,
+        }
+    }
+
+    pub fn update_state(&self) -> &UpdateState {
+        &self.update_state
+    }
+
+    pub fn chosen_update(&self) -> Option<&UpdateImage> {
+        match &self.chosen_update {
+            Some(update) => Some(&update),
+            None => None,
+        }
+    }
+
+    pub fn staging_partition(&self) -> Option<&StagedImage> {
+        match &self.staging_partition {
+            Some(partition_info) => Some(&partition_info),
+            None => None,
+        }
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    /// Transition update state to Idle if the 'chosen' update version is not available
+    pub fn to_idle(&mut self) -> Result<()> {
```

Comment:
I'm thinking that we might want to parcel out the legal transitions and operations themselves to handle them in a more focused fashion.

There have been a handful of discussions on blogs and such that use the type system of Rust to generate a "state machine" of types (even a [macro-ize `crate`](https://github.com/rust-bakery/machine) came of it!). It might be worth using richer types to enforce the allowable transitions internally.

Regardless, I would very much like to see the allowable state transition handling in its own type or Trait impl.

Reply:
I explored the trait option early on and realized (or at least it seemed to me from the reading I did) that `serde` does not play nice when de-serializing into generic types. For the sake of simplicity I opted for runtime checks here.

Reply:
Another thing is that since these states don't internally store any information nor do they act upon anything upon transitions, the use of types and traits to represent these states are somewhat unnecessary (aside from the compile time checks for transition validity, that's always nice).

Reply:
> since I can't tell the compiler the concrete type (state) the status should be de serialized as

One way to do this actually is to have a separate enum in the status struct that indicates what type (state) the rest of the struct should be deserialized as. But then at the point I would have  to have a enum to represent the state already and the I don't really get much from the benefits from having the state as separate types.

Reply:
https://github.com/rust-lang/rfcs/pull/2593
^ This would be really nice to have for this purpose

Reply:
> One way to do this actually is to have a separate enum in the status struct that indicates what type (state) the rest of the struct should be deserialized as. But then at the point I would have to have a enum to represent the state already and I don't really get much from the benefits of having the state as separate types.

Yeah, you could lift in and out of a single enum with all states and use a `get_state_object(&self) -> impl StateMachinery` to read out the "smart transition". But as you said, you'd have to duplicate and that's no bueno. The types themselves would still hold the constraints so that'll still an improvement..

Serde does let you do tagged enums, but you're right about serde not lending a hand w.r.t. nicely deserializing into Trait objects (I wanted the same thing a while back). Which is too bad.

I wish this was nicer to have, we'll probably just have to punt on this until we get something more unwieldy. I still strongly recommend pulling the transition logic into its own type separate from the rest of the driving code. It could be embedded into the serialized data (and even flattened for other consumers if needed).